### PR TITLE
feat: add basic prepare hook

### PR DESCRIPTION
**Pull request by torresxb1**
#### Why is this change necessary?
Adds basic prepare hook structure -- handles zip lambda functions. 
- Created a prepare hook module which is called from the terraform hook entry points
  - Runs terraform commands to initialize application and get json output of terraform plan
  - Uses json output to generate metadata file (translate to cloudformation)
  - maps s3 objects to functions (`Code`) if they have matching bucket and key

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/hook/hook_wrapper.py [hunk 1]
```diff
@@ -80,11 +80,9 @@ def prepare(
             "IACProjectPath": iac_project_path if iac_project_path else str(Path.cwd()),
             "OutputDirPath": output_dir_path,
             "Debug": debug,
+            "Profile": aws_profile,
+            "Region": aws_region,
```
Comment (torresxb1) [inline samcli/lib/hook/hook_wrapper.py:84]: This change isn't necessary. It was needed when I had some dataclasses that I ended up removing. However, I wonder if there's much difference ( e.g. `params.get("Profile")` would be `None` in either case anyway)

File: tests/unit/lib/hook/test_hook_wrapper.py [hunk 2]
```diff
@@ -1,11 +1,7 @@
 """Test Hook Warpper"""
-from asyncio import subprocess
-import json
-from copy import deepcopy
 from pathlib import Path
-from sys import stderr, stdin
 from unittest import TestCase
-from unittest.mock import patch, MagicMock, mock_open, Mock
```
Comment (torresxb1) [inline tests/unit/lib/hook/test_hook_wrapper.py:8]: I believe they were unused

File: samcli/hook_packages/terraform/main.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
...
-Module for Terraform hooks
+Module for Terraform hook entry points
 """
+from .hooks import TerraformHooks
 
 
-def prepare(self, options: dict) -> dict:
-    pass
+def prepare(params: dict) -> dict:
+    tf_hooks = TerraformHooks()
```
Comment (torresxb1) [inline samcli/hook_packages/terraform/main.py:8]: should TerraformHooks be a module instead of a class? I feel like we don't use much functionality of classes (aside from creating a Hooks abstract class and implementing that)

File: samcli/hook_packages/terraform/hooks.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+                os.mkdir(output_dir_path)
+            metadataFilePath = os.path.join(output_dir_path, "template.json")
+            with open(metadataFilePath, "w+") as metadata_file:
+                json.dump(cfn_dict, metadata_file)
+
+            return {"iac_applications": {"MainApplication": {"metadata_file": metadataFilePath}}}
+
+        except (CalledProcessError, OSError) as e:
+            # one of the subprocess.run calls resulted in non-zero exit code or some OS error
+            raise PrepareHookException("There was an error while preparing the Terraform application.") from e
```
Comment (torresxb1) [inline samcli/hook_packages/terraform/hooks.py:344]: @hawflau and I briefly discussed two different ways of handling errors:
```
except SomeInternalException as e:
    raise ExceptionUnderstoodByWarpper(e)
```
```
except SomeInternalException as e:
    return {
        "status": "ERROR",
        "error_message": str(e)
    }
```

We wanted to bring this for discussion in this PR

File: samcli/hook_packages/terraform/hooks.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+from samcli.lib.utils.hash import str_checksum
+from samcli.lib.utils.resources import (
+    AWS_LAMBDA_FUNCTION as CFN_AWS_LAMBDA_FUNCTION,
+)
+
+TF_AWS_LAMBDA_FUNCTION = "aws_lambda_function"
+SUPPORTED_RESOURCE_TYPES = [TF_AWS_LAMBDA_FUNCTION]
+PROVIDER_NAME = "registry.terraform.io/hashicorp/aws"
+
+# max logical id len is 255
```
Comment (torresxb1) [inline samcli/hook_packages/terraform/hooks.py:24]: not sure if this is true, got it from CDK (where the logic for building the logical ID was partially inspired from): https://github.com/aws/aws-cdk/blob/3a7975c35aed53b94d0ea5577e41ad70a81e1390/packages/%40aws-cdk/core/lib/private/uniqueid.ts#L21

File: samcli/hook_packages/terraform/hooks.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    tf_cfn_prop_names = [("filename", "ZipFile"), ("s3_bucket", "S3Bucket"), ("s3_key", "S3Key")]
+    for tf_prop_name, cfn_prop_name in tf_cfn_prop_names:
+        tf_prop_value = tf_properties.get(tf_prop_name)
+        if tf_prop_value is not None:
+            code[cfn_prop_name] = tf_prop_value
+    return code
+
+
+AWS_LAMBDA_FUNCTION_PROPERTY_BUILDER_MAPPING: PropertyBuilderMapping = {
+    "FunctionName": _get_property_extractor("function_name"),
```
Comment (torresxb1) [inline samcli/hook_packages/terraform/hooks.py:196]: these are like one-to-one mapping

File: samcli/hook_packages/terraform/hooks.py (hunk 5 above)
Comment (hawflau) [reply samcli/hook_packages/terraform/hooks.py:23]: I believe the limitation comes from CFN

File: tests/unit/hook_packages/terraform/test_hooks.py [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+    TerraformHooks,
+)
+from samcli.lib.hook.exceptions import PrepareHookException
+from samcli.lib.utils.resources import (
+    AWS_LAMBDA_FUNCTION as CFN_AWS_LAMBDA_FUNCTION,
+)
+
+
+class TestPrepareHook(TestCase):
+    def setUp(self) -> None:
```
Comment (moelasmar) [inline tests/unit/hook_packages/terraform/test_hooks.py:27]: can you add some test cases for the TF applications with Lambda Function, and the aws_s3_object resources.

**Review by moelasmar — CHANGES_REQUESTED** [review]

File: tests/unit/hook_packages/terraform/test_hooks.py [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+            "Layers": ["layer_arn1", "layer_arn2"],
+        }
+
+        self.tf_zip_function_properties: dict = {
+            **self.tf_function_common_properties,
+            "filename": "file.zip",
+        }
+        self.expected_cfn_zip_function_properties: dict = {
+            **self.expected_cfn_function_common_properties,
+            "Code": {"ZipFile": "file.zip"},
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply tests/unit/hook_packages/terraform/test_hooks.py]: done

File: tests/unit/hook_packages/terraform/test_hooks.py [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+
+
+class TestPrepareHook(TestCase):
+    def setUp(self) -> None:
+        self.mock_logical_id_hash = "12AB34CD"
+
+        self.tf_function_common_properties: dict = {
+            "function_name": "myfunc",
+            "architectures": ["x86_64"],
+            "environment": {"variables": {"foo": "bar", "hello": "world"}},
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply tests/unit/hook_packages/terraform/test_hooks.py]: done

File: samcli/hook_packages/terraform/hooks.py [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+        dict
+            information of the generated metadata files
+        """
+        try:
+            output_dir_path = params.get("OutputDirPath")
+            if not output_dir_path:
+                raise PrepareHookException("OutputDirPath was not supplied")
+
+            # initialize terraform application
+            run(["terraform", "init", "-upgrade"], check=True, capture_output=True)
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py]: done

File: samcli/hook_packages/terraform/hooks.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+    for _, resource in cfn_resources.items():
+        if resource.get("Type") == CFN_AWS_LAMBDA_FUNCTION:
+            code = resource.get("Properties").get("Code")
+            bucket = code.get("S3Bucket")
+            key = code.get("S3Key")
+            if bucket and key:
+                obj_hash = _get_s3_object_hash(bucket, key)
+                source = s3_hash_to_source.get(obj_hash)
+                if source:
+                    code["ZipFile"] = source
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py]: done

File: samcli/hook_packages/terraform/hooks.py [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+
+AWS_LAMBDA_FUNCTION_PROPERTY_BUILDER_MAPPING: PropertyBuilderMapping = {
+    "FunctionName": _get_property_extractor("function_name"),
+    "Architectures": _get_property_extractor("architectures"),
+    "Environment": _build_lambda_function_environment_property,
+    "Code": _build_lambda_function_code_property,
+    "Handler": _get_property_extractor("handler"),
+    "PackageType": _get_property_extractor("package_type"),
+    "Runtime": _get_property_extractor("runtime"),
+    "Layers": _get_property_extractor("layers"),
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py:265]: done

File: samcli/hook_packages/terraform/hooks.py [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+    tf_properties: dict
+        Properties of the terraform AWS Lambda function resource
+
+    Returns
+    -------
+    dict
+        The built Code property of a CloudFormation AWS Lambda Function resource
+    """
+    code = {}
+    tf_cfn_prop_names = [("filename", "ZipFile"), ("s3_bucket", "S3Bucket"), ("s3_key", "S3Key")]
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py]: done

File: samcli/hook_packages/terraform/hooks.py [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+
+    Returns
+    -------
+    dict
+        The built Environment property of a CloudFormation AWS Lambda Function resource
+    """
+    environment = tf_properties.get("environment")
+    if not environment:
+        return None
+    return {"Variables": environment.get("variables")}
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py]: done

File: samcli/hook_packages/terraform/main.py (hunk 3 above)
Comment (torresxb1) [reply samcli/hook_packages/terraform/main.py]: changed to module

File: samcli/hook_packages/terraform/hooks.py [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+
+            # Add resource to cfn dict
+            cfn_dict["Resources"][logical_id] = {
+                "Type": resource_translator.cfn_name,
+                "Properties": translated_properties,
+                "Metadata": {"SamResourceId": resource_address, "SkipBuild": True},
+            }
+
+    # map s3 object sources to respective functions
+    _map_s3_sources_to_functions(s3_hash_to_source, cfn_dict["Resources"])
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py:157]: done

File: samcli/hook_packages/terraform/hooks.py [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+            # store S3 sources
+            if resource_type == "aws_s3_object":
+                obj_hash = _get_s3_object_hash(resource_values.get("bucket"), resource_values.get("key"))
+                s3_hash_to_source[obj_hash] = resource_values.get("source")
+
+            resource_translator = RESOURCE_TRANSLATOR_MAPPING.get(resource_type)
+            # resource type not supported
+            if not resource_translator:
+                continue
+            # translate TF resource "values" to CFN properties
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py:139]: done

File: tests/unit/hook_packages/terraform/test_hooks.py [hunk 17]
```diff
@@ -482,36 +483,44 @@ def test_translate_resource_function_with_missing_or_none_properties(self):
         )
         self.assertEqual(translated_cfn_properties, self.expected_cfn_function_properties_with_missing_or_none)
 
-    @patch("samcli.hook_packages.terraform.hooks.hashlib")
-    def test_map_s3_sources_to_functions(self, mock_hashlib):
-        mock_md5 = Mock()
-        mock_hashlib.md5.return_value = mock_md5
-        mock_md5.hexdigest.side_effect = ["hash1", "hash2"]
+    @patch("samcli.hook_packages.terraform.hooks._get_s3_object_hash")
+    def test_map_s3_sources_to_functions(self, mock_get_s3_object_hash):
```
Comment (torresxb1) [inline tests/unit/hook_packages/terraform/test_hooks.py:489]: The changes to this test were supposed to go into a separate commit, sorry 😅

File: samcli/hook_packages/terraform/hooks.py [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+    -------
+    dict
+        information of the generated metadata files
+    """
+    output_dir_path = params.get("OutputDirPath")
+    if not output_dir_path:
+        raise PrepareHookException("OutputDirPath was not supplied")
+
+    try:
+        # initialize terraform application
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks.py:58]: can you add debugging message instead of these comments or adding to it :)

Comment (moelasmar) [reply samcli/hook_packages/terraform/hooks.py:58]: my point is TF commands usually take time, so may be it is something good to let the customers what we are doing right now, so they do not think that there is an issue. Also, if possible to write the commands output and error streams for debugging, and specially in case of failures

File: samcli/hook_packages/terraform/hooks.py [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+    CloudFormation AWS Lambda Function resources
+
+    Parameters
+    ----------
+    s3_hash_to_source: Dict[str, str]
+        Mapping of S3 object hash to S3 object source
+    cfn_resources: dict
+        CloudFormation resources
+    """
+    for _, resource in cfn_resources.items():
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks.py:345]: can we add some debugging message here to let us know that for a lambda resource X, we replaced the S3 properties with some S3 Object file path?

File: tests/unit/hook_packages/terraform/test_hooks.py (hunk 7 above)
Comment (moelasmar) [reply tests/unit/hook_packages/terraform/test_hooks.py:28]: I am asking for a test case to test `_translate_to_cfn` function with a tf plan JSON that contains one Lambda function resource that contains S3 bucket, and S3 Key that are used in another S3_Object resource. I believe `test_translate_to_cfn_with_mapping_s3_source_to_function` is testing tf plan json that contains two lambda resources, one with S3 attributes, and one with filename attribute

**Review by moelasmar — CHANGES_REQUESTED** [review]

File: tests/unit/hook_packages/terraform/test_hooks.py (hunk 7 above)
Comment (torresxb1) [reply tests/unit/hook_packages/terraform/test_hooks.py]: you were right, that test was supposed to test translating with mapping sources but I missed adding the s3 objects in the test data. Fixed now 🙏

**Comment by torresxb1** [comment]
Addressed some feedback, but logging is still pending

File: samcli/hook_packages/terraform/hooks.py (hunk 18 above)
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py]: done

File: samcli/hook_packages/terraform/hooks.py (hunk 19 above)
Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks.py]: done

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 20]
```diff
@@ -56,29 +56,42 @@ def prepare(params: dict) -> dict:
...
 
         # get json output of terraform plan
+        LOG.info("Creating terraform plan and getting JSON output")
         with NamedTemporaryFile() as temp_file:
             run(["terraform", "plan", "-out", temp_file.name], check=True, capture_output=True)
             result = run(["terraform", "show", "-json", temp_file.name], check=True, capture_output=True)
         tf_json = json.loads(result.stdout)
 
         # convert terraform to cloudformation
+        LOG.info("Converting Terraform to CloudFormation")
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:70]: can we say generating metadata file instead of cloud formation converting

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 21]
```diff
@@ -56,29 +56,42 @@ def prepare(params: dict) -> dict:
...
-    except (CalledProcessError, OSError) as e:
+    except CalledProcessError as e:
         # one of the subprocess.run calls resulted in non-zero exit code or some OS error
+        LOG.debug(
+            "Error running terraform command: \n" "cmd: %s \n" "stdout: %s \n" "stderr: %s \n",
+            e.cmd,
+            e.stdout,
+            e.stderr,
+        )
         raise PrepareHookException("There was an error while preparing the Terraform application.") from e
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:92]: can we add unit test cases to cover the TF commands failures?

Comment (torresxb1) [reply samcli/hook_packages/terraform/hooks/prepare.py:92]: there's `test_prepare_with_called_process_error`, or where you thinking of some other tests?

Comment (moelasmar) [reply samcli/hook_packages/terraform/hooks/prepare.py:91]: sorry it is my bad, I missed it .. thanks Rup

**Review by moelasmar — APPROVED** [review]

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+        metadataFilePath = os.path.join(output_dir_path, "template.json")
+        LOG.info("Finished generating metadata file. Storing in %s", metadataFilePath)
+        with open(metadataFilePath, "w+") as metadata_file:
+            json.dump(cfn_dict, metadata_file)
+
+        return {"iac_applications": {"MainApplication": {"metadata_file": metadataFilePath}}}
+
+    except CalledProcessError as e:
+        # one of the subprocess.run calls resulted in non-zero exit code or some OS error
+        LOG.debug(
```
Comment (sriram-mv) [inline samcli/hook_packages/terraform/hooks/prepare.py:85]: Do we need to verify if the terraform binary exists? and prompt to install if it does not?

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+    "Layers": _get_property_extractor("layers"),
+    "Timeout": _get_property_extractor("timeout"),
+}
+
+RESOURCE_TRANSLATOR_MAPPING: Dict[str, ResourceTranslator] = {
+    TF_AWS_LAMBDA_FUNCTION: ResourceTranslator(CFN_AWS_LAMBDA_FUNCTION, AWS_LAMBDA_FUNCTION_PROPERTY_BUILDER_MAPPING),
+}
+
+
+def _build_cfn_logical_id(tf_address: str) -> str:
```
Comment (sriram-mv) [inline samcli/hook_packages/terraform/hooks/prepare.py:287]: Let's add a comment to mark this function as critical, Also add a reference to a cfn doc on logical ids if you can.

changing IDs in the future can cause unintentional re-deployments.

Comment (moelasmar) [reply samcli/hook_packages/terraform/hooks/prepare.py:287]: we will not deploy the generated Metadata file (CFN template) .. for IaCs support, we are not going to support `sam deploy` and `sam package` commands. Also, in case of `sam sync` where we need to deploy the underling application we will use the `terraform deploy` to deploy the application (in other IaCs, we are going to use the IaC tools to deploy, we will not deploy our own generated templates)

**Review by sriram-mv — APPROVED** [review]
Approving with comments.
